pkg/cri/resource-manager/policy/builtin/memtier/pod-preferences.go

```diff
@@ -217,62 +325,123 @@ func coldStartPreference(pod cache.Pod, container cache.Container) (ColdStartPre
+	//            - shared preference explicitly annotated/configured true:
+	//              => shared cores
+	//          b. non-fractional allocation:
+	//            - shared preference explicitly annotated true:
+	//              => shared cores
+	//            - isolated default preference false or explicitly annotated false:
+	//              => exclusive cores
+	//            - isolated default preference true or explicitly annotated true:
+	//              => exclusive cores, prefer isolated
+	//      - 2 full cores <= CPU request
+	//          a. fractional alloction:
+	//            - shared preference explicitly annotated false:
+	//              => mixed cores, prefer isolated only if explicitly annotated
+	//            - otherwise (no shared annotation):
+	//              => shared cores
+	//          b. non-fractional allocation:
+	//            - shared preference explicitly annotated true:
+	//              => shared cores
+	//            - otherwise (no shared annotation):
+	//              => exclusive cores, prefer isolated only if explicitly annotated
+	//
+
+	namespace := container.GetNamespace()
+	request := container.GetResourceRequirements().Requests[corev1.ResourceCPU]
+	qosClass := pod.GetQOSClass()
+	fraction := int(request.MilliValue())
+
+	// easy cases: kube-system namespace, Burstable or BestEffort QoS class containers
+	switch {
+	case namespace == metav1.NamespaceSystem:
+		return 0, fraction, false, cpuReserved
+	case qosClass == corev1.PodQOSBurstable:
+		return 0, fraction, false, cpuNormal
+	case qosClass == corev1.PodQOSBestEffort:
+		return 0, 0, false, cpuNormal
 	}
 
-	qos := pod.GetQOSClass()
+	// complex case: Guaranteed QoS class containers
+	cores := fraction / 1000
+	fraction = fraction % 1000
+	preferIsolated, explicitIsolated := isolatedCPUsPreference(pod, container)
+	preferShared, explicitShared := sharedCPUsPreference(pod, container)
 
-	preferIsol, explicit := podIsolationPreference(pod, container)
-	preferShared, elevate := podSharedCPUPreference(pod, container)
-
-	full, fraction, isolate := 0, 0, false
 	switch {
-	case container.GetNamespace() == metav1.NamespaceSystem:
-		full, fraction = 0, int(req.MilliValue())
-
-	case qos == corev1.PodQOSBurstable || preferShared:
-		full, fraction = 0, int(req.MilliValue())
-
-	case qos == corev1.PodQOSGuaranteed:
-		full = int(req.MilliValue()) / 1000
-		fraction = int(req.MilliValue()) % 1000
-	}
-
-	if !preferShared {
-		switch {
-		case full == 1:
-			if explicit {
-				isolate = preferIsol
-			} else {
-				isolate = true
+	// sub-core CPU request
+	case cores == 0 && fraction < 1000:
+		return 0, fraction, false, cpuNormal
+		// 1 <= CPU request < 2
+	case cores < 2:
+		// fractional allocation, potentially mixed
+		if fraction > 0 {
+			if preferShared {
+				return 0, 1000*cores + fraction, false, cpuNormal
 			}
-		case full > 1:
-			isolate = preferIsol && explicit
+			return cores, fraction, preferIsolated, cpuNormal
```

Comment:
This may be just because of lacking some background information, but I find it somewhat difficult to understand/explain/document why scheduling on isolated cores requires explicit annotation when requesting 2100m CPU but not when requesting 1900m CPU.

Do you think we could simplify the behavior by using `preferIsolated && explicitIsolated` in every case, or am I missing some important use case?

Reply:
@klihub can correct me if I'm wrong, but, I guess the thinking is that allocating exactly one isolated core is "safe" from the workload scheduling point of view. When/if allocating more than one isolated cores the workload must be "isolcpus aware" and handle scheduling/cpu affinity of its threads by itself. In the case of one core this is not a problem as all threads will be forced on this one core

Reply:
Yes, it's exactly as @marquiz said. And that's why we don't assume all workloads to be safe for multi-core mixed allocations but require them to be annotated so.

Reply:
K. Agree with @askervin that this should be documented. A comment in the source code would probably be in order

Reply:
I added a rationale section to the comments explaining the reasons behind the choices when to allocate
isolated vs. ordinary cores in the various cases.
Fixed a comparison thinko in the comments and a typo.
Removed a useless ` + 1000 * fraction` from a `return` on a branch where `fraction` was guaranteed to
always be `0`.